From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sun, 25 Dec 2022 16:55:45 +0100
Subject: [PATCH] Load portal destination chunk before entity teleport

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Gale - https://galemc.org

This patch is based on the following patch:
"Load chunks when entities go through an end portal"
By: PureGero <puregero@gmail.com>
As part of: MultiPaper (https://github.com/MultiPaper/MultiPaper)
Licensed under: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index f55019443b3f33dc4e0f845e5906078b1257175a..e748a67d9ba349574462a7fb27a5dca69a571dc9 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4005,7 +4005,15 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
             if (entity == null) {
                 return false;
+            // Gale start - MultiPaper - load portal destination chunk before entity teleport
+            } else {
+                if (entity.level.galeConfig().gameplayMechanics.technical.loadPortalDestinationChunkBeforeEntityTeleport) {
+                    BlockPos pos = BlockPos.containing(position);
+                    world.getChunkSource().addRegionTicket(TicketType.PORTAL, new ChunkPos(pos), 1, pos);
+                    world.getChunkAt(pos);
+                }
             }
+            // Gale end - MultiPaper - load portal destination chunk before entity teleport
 
             entity.restoreFrom(this);
             entity.moveTo(destX, destY, destZ, yaw, f2);
diff --git a/src/main/java/org/galemc/gale/configuration/GaleWorldConfiguration.java b/src/main/java/org/galemc/gale/configuration/GaleWorldConfiguration.java
index 38488750f556ebae7ed41a24b88f5752f3b54ec7..4c9bb806aa6509894de7f09b633c0c3d74b17233 100644
--- a/src/main/java/org/galemc/gale/configuration/GaleWorldConfiguration.java
+++ b/src/main/java/org/galemc/gale/configuration/GaleWorldConfiguration.java
@@ -124,6 +124,11 @@ public class GaleWorldConfiguration extends ConfigurationPart {
 
         }
 
+        public Technical technical;
+        public class Technical extends ConfigurationPart {
+            public boolean loadPortalDestinationChunkBeforeEntityTeleport = false; // Gale - MultiPaper - load portal destination chunk before entity teleport
+        }
+
         public boolean arrowMovementResetsDespawnCounter = true; // Gale - Purpur - make arrow movement resetting despawn counter configurable
         public double entityWakeUpDurationRatioStandardDeviation = 0.2; // Gale - variable entity wake-up duration
         public boolean hideFlamesOnEntitiesWithFireResistance = false; // Gale - Slice - hide flames on entities with fire resistance
